File: ts/src/instructions/sunsetRiskProfile.ts
import { PublicKey, TransactionInstruction } from '@solana/web3.js';

import { globalConfigPda, globalVaultPda } from '../pdas.js';
import { ro, rw, signerRw, ydeltaIx } from './_helpers.js';
import { Writer } from './_serialise.js';
import { InstructionTag } from './_tags.js';

/**
 * Tag 38 — `SunsetRiskProfile`. Vault-admin-gated. Flips
 * `profile.is_sunset = 1`, beginning the wind-down lifecycle:
 *   - deposits, new orders, order updates, and matches all reject with
 *     `VaultProfileSunset` (code 54)
 *   - withdrawals, curator-fee claims, cancels, repay, settle, and
 *     liquidate continue to operate
 *
 * Reversible via `resumeRiskProfileInstruction` (tag 39).
 * `removeRiskProfileInstruction` (tag 37) and
 * `adminCancelRiskProfileOrderInstruction` (tag 40) both gate on this flag.
 */
export interface SunsetRiskProfileArgs {
  payer: PublicKey;
  mint: PublicKey;
  profileId: number;
}

export function sunsetRiskProfileInstruction(
  args: SunsetRiskProfileArgs,
): TransactionInstruction {
  const vault = globalVaultPda(args.mint)[0];
  const data = new Writer()
    .u8(InstructionTag.SunsetRiskProfile)
    .u8(args.profileId)
    .toBuffer();
  return ydeltaIx(
    [signerRw(args.payer), ro(globalConfigPda()[0]), rw(vault)],
    data,
  );
}

File: programs/ydelta/src/program/processor/sunset_risk_profile.rs
//! `SunsetRiskProfile` — vault-admin sunset for a risk profile. Signer is
//! the vault admin (via `SunsetRiskProfileContext`). Sets `is_sunset = 1`,
//! which blocks deposits, new orders, order updates, and matches; withdrawals,
//! cancels, repay, settle, and liquidate stay open during wind-down.
//! `AdminCancelRiskProfileOrder` and `RemoveRiskProfile` both gate on this
//! flag.

use std::cell::RefMut;

use borsh::{BorshDeserialize, BorshSerialize};
use hypertree::{HyperTreeReadOperations, NIL};
use solana_program::{account_info::AccountInfo, entrypoint::ProgramResult, pubkey::Pubkey};

use crate::program::YdeltaError;
use crate::require;
use crate::state::vault::{
    get_mut_helper_risk_profile, GlobalVaultFixed, RiskProfile, RiskProfileTreeReadOnly,
};
use crate::state::GLOBAL_VAULT_FIXED_SIZE;
use crate::validation::loaders::SunsetRiskProfileContext;

/// Sunset parameters.
#[derive(BorshDeserialize, BorshSerialize, Clone, Copy)]
pub struct SunsetRiskProfileParams {
    /// Risk profile ID to sunset (1-based; 0 is the sentinel).
    pub profile_id: u8,
}

/// Set the profile's `is_sunset` flag, blocking deposits / new orders /
/// updates / matches while keeping withdrawals and close-outs open.
pub fn process_sunset_risk_profile(
    _program_id: &Pubkey,
    accounts: &[AccountInfo],
    data: &[u8],
) -> ProgramResult {
    let params = SunsetRiskProfileParams::try_from_slice(data)?;
    let SunsetRiskProfileContext { _payer: _, vault } = SunsetRiskProfileContext::load(accounts)?;

    let data: &mut RefMut<&mut [u8]> = &mut vault.info.try_borrow_mut_data()?;
    let (fixed_bytes, dynamic) = data.split_at_mut(GLOBAL_VAULT_FIXED_SIZE);
    let header: &mut GlobalVaultFixed = bytemuck::from_bytes_mut(fixed_bytes);

    let probe = RiskProfile::new_empty(params.profile_id, Pubkey::default(), 1, 1);
    let idx = {
        let tree = RiskProfileTreeReadOnly::new(dynamic, header.risk_profiles_root_index, NIL);
        tree.lookup_index(&probe)
    };
    require!(
        idx != NIL,
        YdeltaError::VaultProfileNotFound,
        "sunset_risk_profile: profile_id {} not found",
        params.profile_id
    )?;

    let profile = get_mut_helper_risk_profile(dynamic, idx).get_mut_value();
    profile.is_sunset = 1;
    Ok(())
}
